### PR TITLE
Add alias name

### LINKED ISSUE #1109: add texterAliasfirstName as an option for sending
Add an option for folks, especially women, to add a 'texterAliasfirstName' to their user record to optionally send as - like 'Frank' (alias) rather than 'Lydia' (actual First Name). Often, texters IDing as women receive harassing replies back, perhaps as a reaction to common 'gender association' of their first name.  
Would also necessitate a preferredTexterFirstName field to use for actual  message creation (like {preferredTexterfirstName} token), with user firstName as default, but alias as an option.

### PR DESCRIPTION
**Pull request by filafb**
# Fixes #1109 

## Description

This feature allows user to send messages using an alias. If there is no alias for that texter, the message will include texter's first name when the interactions is _{ texterAliasOrFirstName }_

I had to change the resolver implementation for texter [here](https://github.com/filafb/Spoke/blob/63cfbac5bc423bfe09fb354d17df26f291069f10/src/server/api/assignment.js#L116). The previous implementation wasn't bringing me a new instance for texter after adding/changing an user alias. Not sure if it was a caching strategy, but now I'm finding the updated user instance direct from the db. 

I also fixed an issue (might be the issue #1358) that was avoiding the Dialog Box Edit User to be populated with user data. Not when user is editing a person data, this dialog box brings their current information.

# Checklist:

- [x] I have manually tested my changes on desktop and mobile
- [x] The test suite passes locally with my changes
- [x] If my change is a UI change, I have attached a screenshot to the description section of this pull request
- [x] [My change is 300 lines of code or less](https://github.com/MoveOnOrg/Spoke/blob/main/CONTRIBUTING.md#your-first-code-contribution), or has a documented reason in the description why it’s longer
- [ ] I have made any necessary changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
_* Added test for the new DB schema. I don't have much experience testing GraphQL. Happy to follow directions if anybody wants to point it to me._ 
- [ ] My PR is labeled [WIP] if it is in progress

interactions
list of people
edit user

### CODE REVIEW

**Comment by arena** [comment]
🎉🎉🎉

**Comment by joemcl** [comment]
thanks @filafb and @ibrand , and I hope I can make time to review this tonight. cc: @deasterdaywfp @lperson @bdatkins @suegibson .

File: __test__/backend.test.js
Comment (bdatkins) [inline __test__/backend.test.js]: General comment: Can you please use the prettier defaults (such as double quotes) so it's easier to see what has changed?

File: src/server/api/user.js
Comment (bdatkins) [inline src/server/api/user.js:180]: Should this be first name or alias in this case?

**Review by bdatkins — CHANGES_REQUESTED** [review]
Looks good!  A really great addition to Spoke. Just a couple minor points from me.

File: src/containers/UserEdit.jsx
Comment (schuyler1d) [inline src/containers/UserEdit.jsx]: Interested in thoughts from others -- I feel like "Alias" doesn't have much context especially when the user is just creating their account.  Maybe label (or if there's a hint/description property) could be "Alias for texting" -- and make it clear someone can leave it blank?

File: src/containers/UserEdit.jsx
Comment (schuyler1d) [inline src/containers/UserEdit.jsx]: what/where is this query property?  the query result will end up in the `data` property above (I believe).  I think this is unnecessary.

File: src/server/api/assignment.js
Comment (schuyler1d) [inline src/server/api/assignment.js]: Yes, please revert this -- we definitely want this data to come from the cache.
The issue you observed IS something we should fix.  If you want to do it in this PR, then it looks like `editUser()` correctly calls ` await cacheableData.user.clearUser(member.id, member.auth0_id);`
The problem is that clearUser doesn't currently clear the loaders (local memory) cache.  What you want to do is copy the pattern found in:
src/server/models/cacheable_queries/organization.js (or campaign.js)
where you import `loaders` from "../../models" and then at the end of clearUser (in user.js in the same dir) you run `loaders.user.clear(Number(id))` and do the same thing with type string (to be extra sure).

Caching is a super-crucial part of scaling Spoke as it's a very 'chatty' client which makes the database the bottleneck, so it's important to retain (and in the future extend) the data that's cached -- especially for anything texter-related.

File: src/server/api/schema.js
Comment (schuyler1d) [inline src/server/api/schema.js]: Since alias can easily override firstName -- should we chomp off whitespace before and after (probably good to do for first/last name, if you want, too)

File: src/server/models/user.js
Comment (schuyler1d) [inline src/server/models/user.js]: I think the changes in this file are lint-churn.  `prettier` really prefers double-quotes here, so at least please make sure you're using the githook in package.json or running `prettier --write` on changed files on your own.

**Review by schuyler1d — CHANGES_REQUESTED** [review]
This is a great complete implementation of the alias feature!
I've requested some minor changes here or there, but I think this is very close.
Thanks!

File: src/containers/UserEdit.jsx
Comment (arena) [reply src/containers/UserEdit.jsx]: In my mockup I believe I used "Texting Alias"

**Comment by arena** [comment]
Not sure if this is relevant, but this shows the form field with a label of 
"Texing Alias (optional)"
https://github.com/MoveOnOrg/Spoke/issues/1357

File: __test__/backend.test.js
Comment (filafb) [reply __test__/backend.test.js]: Sure! Will do that. For some reason, husky & lint-staged are not running for me when I commit my stuff.

File: src/server/api/user.js
Comment (filafb) [reply src/server/api/user.js:180]: The code is doing something like `FirstName (alias) lasName` when alias is provided. If not, it will display `firstName lasName`. You can see that on this [print screen](https://user-images.githubusercontent.com/38737958/72227990-9e7fb400-3570-11ea-8617-049efca1f045.png)

File: src/containers/UserEdit.jsx
Comment (filafb) [reply src/containers/UserEdit.jsx]: Is everybody fine if I change it to `Texting Alias`?

File: src/containers/UserEdit.jsx
Comment (filafb) [reply src/containers/UserEdit.jsx]: I think you are right. Don't think we are using it inside the component. Will review it.

File: src/server/api/assignment.js
Comment (filafb) [reply src/server/api/assignment.js]: Alright. Will take a look at it.

File: src/server/models/user.js
Comment (filafb) [reply src/server/models/user.js]: Will do it!

File: src/server/api/schema.js
Comment (filafb) [reply src/server/api/schema.js]: Sorry, I didn't fully understand the suggestion here.

**Comment by filafb** [comment]
Thanks for all comments people! I will probably work on them this weekend! @schuyler1d @bdatkins

File: src/containers/UserEdit.jsx
Comment (arena) [reply src/containers/UserEdit.jsx]: That is most intuitive
"Texting Alias (optional)"
Nobody has voiced any opposition and it's nice and descriptive, especially with the "(optional)" added to it. 

Yay!!!

**Review by bdatkins — DISMISSED** [review]
Nice work!

**Comment by filafb** [comment]
Hey @schuyler1d and @bdatkins! Just pushed new code that fixes the stuff you have commented on.

**Comment by arena** [comment]
🙌🎉✨🎉🎉❤️🎉🎉✨🎉✨

**Comment by ibrand** [comment]
@filafb some things I noticed when testing this out on the [heroku deployment of this branch](https://spoke-review-add-aliasn-h6crwu.herokuapp.com/):

1. When I sign up with an alias, it doesn't seem to save that alias vs when I go to my account details and edit it there, it works.

2. For a live campaign, if I change my name in the middle, I'm still texting with my old name

3. If the `texterName` variable was used, it doesn't use my alias, so I have to make sure as an admin to ignore that one and use `texterNameOrAlias`

4. Strangely when i used the `texterName` var it used the first name from my contact list as my name instead. Not sure what to make of that

Thank you so much for working on this! It's going to mean a lot for texters

**Review by ibrand — CHANGES_REQUESTED** [review]
I want to make sure to get answers to the QA questions I pose in my comment before merging. thank you for your work so far!

**Comment by arena** [comment]
not to answer for @filafb ... but I see your questions @ibrand and I have some thoughts
1. I don't think I should be prompted for an alias on sign up... just a way to add one after sign up
2. if you change any part of your name it won't have an effect on an existing assignment... so that seems expected (even if it's surprising to a user)
3. & 4. (over my head, no opinion)

File: src/server/api/schema.js
Comment (schuyler1d) [reply src/server/api/schema.js]: I'm worried about the case where someone accidentally adds `__` (spaces) into the optional Alias field.  Since it overrides first_name in places, I think it's important to either remove front and leading spaces (so that 'only spaces' alias turns into an empty string and will then evaluate to false), and/or test that in this place.

Since we'd be stripping whitespace, I was thinking that might be good to do, even for the fields first/last name, as well, but that has less of a fatal result.

**Comment by filafb** [comment]
Hi @ibrand. Bellow my answers:
1. I will check it. I haven't worked in anything related to the sign up page, but since the option is available, I think it should save the alias.
2. You mean the name or you also tried to change the alias?
3. That is correct. It was part of my question [here](https://github.com/MoveOnOrg/Spoke/issues/1109#issuecomment-533139062). It also gives the admin the power to use texter name when required.
4. I didn't get it. What is the difference between your name and your first name from my contact list?

File: src/server/api/schema.js
Comment (filafb) [reply src/server/api/schema.js]: Got it. Will update the PR, trimming all three fields here.

Comment (filafb) [reply src/server/api/schema.js]: @schuyler1d  Just pushed a new version. Check it out and let me know!

**Comment by filafb** [comment]
@ibrand Just pushed a new version that should fix item number 1 in your list. Check it out.

**Comment by ibrand** [comment]
~@filafb  I'm now seemingly unable to save my alias if I try and change it from the profile screen [in the test deploy](https://spoke-review-add-aliasn-h6crwu.herokuapp.com/)?~ This looks good on my end! Thanks for answering all the questions and for the revision @filafb ! ⭐️

**Review by ibrand — DISMISSED** [review]

File: src/containers/UserEdit.jsx
Comment (schuyler1d) [inline src/containers/UserEdit.jsx]: do we need to test if `this.props.data` is available?  It's possible the component mounts without data available until the network connection is returned.

Can we add a comment here (once we figure out the need).
This needs testing/QA for:
* signup screen
* texter editing themselves screen
* admin editing a texter screen.

I worry that putting this.props.data.currentUser will only work for the second case (and possibly break the others)
Will defer to @ibrand on what we can QA vs. ensure we have a test for, for these use cases.

Comment (filafb) [reply src/containers/UserEdit.jsx]: We can have here a function that returns either `this.props.data.currentUser` or null, if it catches an error reading `this.props.data`. What do you think? @schuyler1d

Comment (filafb) [reply src/containers/UserEdit.jsx]: Ok, I will wait the QA to change it if needed. I was thinking about a `try/catch` block to define what `editUser` will be called with, but that should work too. And if that might be the case (this.props.data being undefined for some cases) I think it's a safe measure. I honestly don't know a good strategy to test it, specially if it is a race condition. Maybe changing the lifecycle method to `componentDidMount` could help?

Comment (schuyler1d) [reply src/containers/UserEdit.jsx]: I think before QA, the logic I mentioned above needs to be in -- or are you arguing that we can leave it just as `props.data.currentUser`? even in the `!props.userId` case?

Comment (filafb) [reply src/containers/UserEdit.jsx]: Oh, my bad! Sorry. I will update it later and do some tests. Thanks!!!

Comment (filafb) [reply src/containers/UserEdit.jsx]: Hey @schuyler1d! Just tested the scenario you listed above and you were right. Editing a user in the list would bring the current logged in user info to Edit Page, not the user I clicked in. I reverted my changes and seems it is working now. Before, when I click to edit a user, the Edit Page was empty, but now it is bringing all the correct info. I think there was another change in the code that fixed that behavior I had seen.

**Review by schuyler1d — DISMISSED** [review]
Ok, I think this is good to go -- this will be so helpful for texters!!!! :tada:

**Comment by arena** [comment]
🎉 🙌

File: migrations/20190916145342_add_user_alias.js
Comment (schuyler1d) [inline migrations/20190916145342_add_user_alias.js]: I just realized we should default this to be null so the schema change is non-locking

**Review by schuyler1d — APPROVED** [review]
:tada:

**Pull request merged**
